File: java/src/main/java/de/activegroup/Animal.java
package de.activegroup;

public sealed interface Animal {

    /*
     * Tier auf dem texanischen Highyway:
     * - Gürteltier      ODER
     * - Klapperschlange
     * Summe
     */

    default public Animal runOver() {
        return switch (this) {
            case Dillo(Liveness l, double w) ->
                new Dillo(Liveness.DEAD, w);
            case Snake(double length, double width) ->
                new Snake(length, 0);
        };
    }

    default public Animal feed(double amount) {
        return switch (this) {
          case Dillo(Liveness liveness, double weight) ->
                  switch (liveness) {
                    case Liveness.ALIVE ->
                            new Dillo(liveness, weight + amount);
                    case Liveness.DEAD ->
                            this;
                  };
          case Snake(double length, double width) ->
              new Snake(length, width + amount);
        };
    }

    /*
     * Gürteltier hat folgende Eigenschaften:
     * - lebendig oder tot        -UND-
     * - Gewicht
     * zusammengesetzte Daten / Produkt
    */
    record Dillo(Liveness liveness, double weight) implements Animal {
        // @Override
        public Dillo runOver() {
            return new Dillo(Liveness.DEAD, weight);
        }
    }

    /*
     * Eine Klapperschlange hat:
     * - Länge  UND
     * - Dicke
     */
    public record Snake(double length, double width) implements Animal {
        // @Override
        public Snake runOver() {
            return new Snake(this.length, 0);
        }
    }

}

File: kotlin/src/main/kotlin/Animal.kt
package de.activegroup

/*
 * Tier auf dem texanischen Highway:
 * - Gürteltier  ODER
 * - Klapperschlange
 * Summe
 *
 * Gürteltier hat folgende Eigenschaften:
 * - lebendig oder tot      -UND-
 * - Gewicht
 * Produkt
 *
 * Klapperschlange hat folgende Eigenschaften:
 * - Länge
 * - Dicke
 */

enum class Liveness {
    ALIVE, DEAD
}

typealias Weight = Double

sealed interface Animal

// Zustand des Gürteltiers zu einem Zeitpunkt
data class Dillo(val liveness: Liveness,
    val weight: Weight): Animal
data class Snake(val length: Double, val width: Double): Animal

// lebendiges Gürteltier, 10kg
val dillo1 = Dillo(Liveness.ALIVE, 10.0)
// totes Gürteltier, 8kg
val dillo2 = Dillo(Liveness.DEAD, 8.0)

// Gürteltier überfahren
fun runOverDillo(dillo: Dillo): Dillo =
    Dillo(Liveness.DEAD, dillo.weight)

fun runOverAnimal(animal: Animal): Animal =
    when (animal) {
        is Dillo -> Dillo(Liveness.DEAD, animal.weight)
        is Snake -> Snake(animal.length, 0.0)
    }